Leave dungeon

gabriellemadden opened the pull request:
# #161 Leave Dungeon Button

## Demo
Here's how our leave dungeon panel appears:
<img width="1433" alt="Screen Shot 2021-12-07 at 3 16 25 PM" src="https://user-images.githubusercontent.com/49839049/145099890-1bcc294a-adb4-4f30-9fce-d54f239940ad.png">

## Manual Testing
### Criteria: The "leave dungeon" button/panel is only accessible when the player is in a dungeon
Cases checked using the knight's training arena:
1. Enter dungeon, see map button change to leave dungeon button -> passed
2. Leave dungeon and see dungeon button change back to map button
  a. By pressing "leave" on the panel in a dungeon -> passed
  b. By getting killed in the dungeon -> passed
  c. By completing a dungeon and using the exit portal -> passed

### Criteria: The leave dungeon panel lets you choose to "stay" or "leave"
1. pressing the leave dungeon icon in the top left GUI opens a dialogue with "leave" and "stay" options -> passed
1. pressing stay closes the panel and leaves the player inside the dungeon -> passed
2. pressing leave teleports the player to the same location that using the exit portal would have -> passed

### Criteria: Player should not be penalized when leaving dungeon (items, glory, stats remain unchanged)
1. glory, stats, items remain unchanged when pressing leave -> passed

## To-do
One question: How is TextDefinitions.json updated? I see catalogues is in the `.gitignore` so if I were to change TextDefinitions I wouldn't be able to include it in my PR. For now, I have english strings hard coded in for the panel tooltip, title, description, and stay button, but I can update these strings to use `Utils.getTextDefs()` when I know what the correct procedure is.

Arcanorum commented:
@gabriellemadden There are some notes about the defining a piece of text in the translations spreadsheet in this guide:
https://github.com/Arcanorum/dungeonz/blob/master/guides/items/ITEMS.md#add-a-translation
This part of the process can be slightly cumbersome. Will review how it works next week but for now it is still fine.
Basically just decide what you want the ID/label of the text to be (left column) then set the actual text value in the English column.

TextDefinitions.json (like any of the "catalogues" as I call them), are generated when the server starts based on some source data input.
The following script takes in the translations spreadsheet file and basically just validates the contents and formats the output to be easier for the client to work with:
https://github.com/Arcanorum/dungeonz/blob/master/server/src/scripts/TextDefinitionsParser.js

gabriellemadden commented:
Alright, I changed the xlsx file to appear as the following, and updated the corresponding code: 

<img width="647" alt="Screen Shot 2021-12-14 at 10 40 15 PM" src="https://user-images.githubusercontent.com/49839049/146120322-5b5ebedd-d522-4f49-af89-822f28e47f4f.png">

Is this sufficient? If so, I'll add the changes to the translations spreadsheet linked in the guide you sent ^^^

Arcanorum commented on client/src/components/game/gui/GUI.jsx:
```diff
@@ -49,6 +51,7 @@ function GUI() {
     const [trackedTask, setTrackedTask] = useState(null);
     const [loggedIn, setLoggedIn] = useState(ApplicationState.loggedIn);
     const [targetDungeonPortal, setTargetDungeonPortal] = useState(null);
+    const [inDungeon, setInDungeon] = useState(null);
```
I think this should use a boolean for the default value instead of `null`.
While `null` is falsy and does work, when it is used it implies that the value can be an object or some other kind of complex value.
Using `false` here instead would be more explicit that this is a yes/no kind of value.

Arcanorum commented on client/src/components/game/gui/panels/leave_dungeon/LeaveDungeonPanel.jsx:
```diff
@@ -0,0 +1,62 @@
+import PanelTemplate from "../panel_template/PanelTemplate";
+import Utils from "../../../../../shared/Utils";
+import leaveDungeonIcon from "../../../../../assets/images/gui/hud/leave-dungeon-icon.png";
+import "leaflet/dist/leaflet.css";
```
Unused import.

Arcanorum commented on client/src/components/game/gui/panels/leave_dungeon/LeaveDungeonPanel.scss:
```diff
@@ -0,0 +1,52 @@
+@import "../../../../../shared/Variables";
+
+
```
1 too many linebreaks.

Arcanorum commented on server/src/WebSocketEvents.js:
```diff
@@ -763,6 +763,20 @@ eventResponses.start_dungeon = (clientSocket, data) => {
+    if (clientSocket.inGame === false) return;
+
+    /** @type {Dungeon} */
+    const dungeon = clientSocket.entity.board.dungeon;
```
A check for whether the board property of the entity isn't null is needed before here, as a dead player in the dungeon would be able to send the "leave_dungeon" event, but when they die their `board` property is nulled, so accessing `dungeon` in `clientSocket.entity.board.dungeon` would be undefined and throw an exception.

This effect can be produced by going into a dungeon, enter `/killself` in chat, then in the browser console, enter `dungeonz.states.ApplicationState.connection.sendEvent("leave_dungeon");` to manually send the leave dungeon event.

gabriellemadden replied:
Would the appropriate behaviour in this case be to return and do nothing if the board is null, or would the event handler need to find some way to spawn the player in the overworld?

Arcanorum replied:
It is fine to just return like this.
It is as edge-case problem that can only be triggered by a user manually sending this event (i.e. through the console), as when their character is in the dead state, the client should only show them the respawn option on the GUI, so they shouldn't normally be able to press the leave dungeon button while dead.

Arcanorum commented on server/src/WebSocketEvents.js:
```diff
@@ -763,6 +763,20 @@ eventResponses.start_dungeon = (clientSocket, data) => {
 
+/**
+ * @param {*} clientSocket
+ * @param {String} data.dungeonID - The ID of a dungeon manager.
```
This line of param docs for "data.dungeonID" isn't used so can be removed, as only `clientSocket` is used in the callback function.

Arcanorum commented:
> Alright, I changed the xlsx file to appear as the following, and updated the corresponding code:
> 
> <img alt="Screen Shot 2021-12-14 at 11 02 48 PM" width="449" src="https://user-images.githubusercontent.com/49839049/146121169-9416fdfa-2db9-48f3-94d7-6a7e8e925ce0.png">
> 
> Is this sufficient? If so, I'll add the changes to the translations spreadsheet linked in the guide you sent ^^^

Yes that looks ok.

Just some minor tweaks needed to some other bits in the PR.

Arcanorum reviewed (approved):
lgtm

gabriellemadden commented:
Okay great, changes have been added to the translations spreadsheet as well. I only put in the English column, although I see there’s a GOOGLETRANSLATE function in google sheets which is pretty nifty. Is this how you populate the other columns normally or do you have someone to verify the translations are correct first?

Arcanorum commented:
Yeah, I usually just copy down the previous row so it includes the Google Translate API formula.
I use it for a basic coverage of each language, but they can be manually edited by anyone if they see that it isn't quite right.
The accuracy of the translations it does for a lot of the languages is a bit of a crapshoot according to several players that know those languages natively, to the point that they just use the English setting anyway lol, so I was thinking of making it be purely filled out by real people, which itself is a large and ever expanding task... :/
